datafusion/physical-plan/src/repartition/distributor_channels.rs

```diff
@@ -155,19 +159,32 @@ impl<T> Clone for DistributionSender<T> {
 
 impl<T> Drop for DistributionSender<T> {
     fn drop(&mut self) {
-        let mut guard_channel = self.channel.lock();
-        guard_channel.n_senders -= 1;
+        let n_senders_pre = self.channel.n_senders.fetch_sub(1, Ordering::SeqCst);
+        if n_senders_pre > 1 {
+            return;
+        }
+
+        let receivers = {
+            let mut state = self.channel.state.lock();
 
-        if guard_channel.n_senders == 0 {
             // Note: the recv_alive check is so that we don't double-clear the status
-            if guard_channel.data.is_empty() && guard_channel.recv_alive {
+            if state
+                .data
+                .as_ref()
+                .map(|data| data.is_empty())
+                .unwrap_or_default()
```

Comment:
Could we put this into ChannelState as a function with a more readable name? Like

```rust
if state.channel_gone() {
...
```